Gaming (#518)
solana-developers/solana-cookbook: docs/gaming/energy-system.md

---
title: Energy System
head:
  - - meta
    - name: title
      content: Solana Cookbook | Energy System
  - - meta
    - name: og:title

>       content: Solana Cookbook | Energy System

Reviewer: ```suggestion
      content: Solana Cookbook | Energy System in Solana Games
```

  - - meta
    - name: description

>       content: Many casual games in traditional gaming use energy systems. This is how you can build it on chain.

Reviewer: ```suggestion
      content: Build Energy Systems for Casual Games on Solana with these Easy Steps
```

  - - meta
    - name: og:description

>       content: Many casual games in traditional gaming use energy systems. This is how you can build it on chain.

Reviewer: ```suggestion
      content: Build Energy Systems for Casual Games on Solana with these Easy Steps
```

  - - meta
    - name: og:image
      content: https://solanacookbook.com/cookbook-sharing-card.png
  - - meta
    - name: og:image:alt
      content: Solana splash card
  - - meta
    - name: twitter:card
      content: summary
  - - meta
    - name: twitter:site
      content: "@solanacookbook"
  - - meta
    - name: twitter:image
      content: "https://solanacookbook.com/cookbook-sharing-card.png"
  - - meta
    - name: robots
      content: index,follow,noodp
  - - meta
    - name: googlebot
      content: index,follow
footer: MIT Licensed
---

# Energy System

> Many casual games in traditional gaming use energy systems. Meaning that actions in the game cost energy which refills over time. This is how you can build it on chain.

Reviewer: ```suggestion
Casual games commonly use energy systems, meaning that actions in the game cost energy which refills over time. In this guide we will walk through how to build one on Solana.
```

> Recommended to start with the [Hello World Example](../gaming/hello-world) if you do not have any prior Solana knowledge.

Reviewer: ```suggestion
It is rcommended to start with the [Hello World Example](../gaming/hello-world) if you do not have any prior Solana knowledge.
```

> Here is a complete example source code based on the Solana dapp scaffold with a react client:

Reviewer: ```suggestion
Here is a complete example source code based on the Solana dapp scaffold with a react client:
```
```suggestion
Here is the complete example source code based on the Solana dapp scaffold with a react client:
```

[Source](https://github.com/Woody4618/solumberjack)<br/>

## Anchor program

> Here we will build a program which refills energy over time which the player can then use to perform actions in the game.

Reviewer: ```suggestion
In this tutorial, we will guide you through the process of creating a program that gradually replenishes the player's energy reserves over time. This, in turn, will enable them to execute various actions within the game.
```

> In our example it will be a lumber jack which chops trees. Every tree will reward one wood and cost one energy.

Reviewer: ```suggestion
In our example, a lumber jack will chops trees with every tree rewarding one wood and cost one energy.
```

### Creating the player account

> First the player needs to create an account which saves the state of our player. Notice the last_login time which will save the current unix time stamp of the player he interacts with the program.

Reviewer: ```suggestion
First the player needs to create an account which saves the state of our player. Notice the last_login time saves the current unix time stamp of the player he interacts with the program.
```

> Like this we will be able to calculate how much energy the player has at a certain point in time.

Reviewer: ```suggestion
With this state, we will be able to calculate how much energy the player has at a certain point in time.
```

> We also have a value for wood which will store the wood the lumber jack chucks in the game.

Reviewer: ```suggestion
We also have a value for how much wood the lumber jack chucks in the game.
```

```rust

pub fn init_player(ctx: Context<InitPlayer>) -> Result<()> {
    ctx.accounts.player.energy = MAX_ENERGY;
    ctx.accounts.player.last_login = Clock::get()?.unix_timestamp;
    Ok(())
}

...

#[derive(Accounts)]
pub struct InitPlayer <'info> {
    #[account(
        init,
        payer = signer,
        space = 1000,
        seeds = [b"player".as_ref(), signer.key().as_ref()],
        bump,
    )]
    pub player: Account<'info, PlayerData>,
    #[account(mut)]
    pub signer: Signer<'info>,
    pub system_program: Program<'info, System>,
}

#[account]
pub struct PlayerData {
    pub name: String,
    pub level: u8,
    pub xp: u64,
    pub wood: u64,
    pub energy: u64,
    pub last_login: i64
}
```

### Choping trees

> Then whenever the player calls the chop_tree instruction we will check if the player has enough energy and reward him with one wood.

Reviewer: ```suggestion
Then whenever the player calls the `chop_tree` instruction we will check if the player has enough energy and reward him with one wood.
```

```rust
    #[error_code]
    pub enum ErrorCode {
        #[msg("Not enough energy")]
        NotEnoughEnergy,
    }

    pub fn chop_tree(mut ctx: Context<ChopTree>) -> Result<()> {
        let account = &mut ctx.accounts;
        update_energy(account)?;

        if ctx.accounts.player.energy == 0 {
            return err!(ErrorCode::NotEnoughEnergy);
        }

        ctx.accounts.player.wood = ctx.accounts.player.wood + 1;
        ctx.accounts.player.energy = ctx.accounts.player.energy - 1;
        msg!("You chopped a tree and got 1 wood. You have {} wood and {} energy left.", ctx.accounts.player.wood, ctx.accounts.player.energy);
        Ok(())
    }
```

### Calculating the energy

> The interesting part happens in the update_energy function. We check how much time has passed and calculate the energy that the player will have at the given time.

Reviewer: ```suggestion
The interesting part happens in the `update_energy` function. We check how much time has passed and calculate the energy that the player will have at the given time.
```

> The same thing we will also do in the client. So we basically lazily update the energy instead of polling it all the time.

Reviewer: ```suggestion
We will do the same in the client. We lazily update the energy instead of polling it all the time.
```

The is a common technique in game development.

```rust

const TIME_TO_REFILL_ENERGY: i64 = 60;
const MAX_ENERGY: u64 = 10;

pub fn update_energy(ctx: &mut ChopTree) -> Result<()> {
    let mut time_passed: i64 = &Clock::get()?.unix_timestamp - &ctx.player.last_login;
    let mut time_spent: i64 = 0;
    while time_passed > TIME_TO_REFILL_ENERGY {
        ctx.player.energy = ctx.player.energy + 1;
        time_passed -= TIME_TO_REFILL_ENERGY;
        time_spent += TIME_TO_REFILL_ENERGY;
        if ctx.player.energy == MAX_ENERGY {
            break;
        }
    }

    if ctx.player.energy >= MAX_ENERGY {
        ctx.player.last_login = Clock::get()?.unix_timestamp;
    } else {
        ctx.player.last_login += time_spent;
    }

    Ok(())
}
```

## Js client

Here is a complete example based on the Solana dapp scaffold with a react client:

> [Source](https://github.com/Woody4618/solumberjack)<br/>

Reviewer: Are these links still right?

Author: updated

### Create connection

In the Anchor.ts file we create a connection:

```js
export const connection = new Connection(
    "https://api.devnet.solana.com",
    "confirmed"
);
```

Notice that the confirmation parameter is set to 'confirmed'. This means that we wait until the transactions are confirmed instead of finalized. This means that we wait until the super majority of the network said that the transaction is valid. This takes around 400ms and there was never a confirmed transaction which did not get finalized. So for games this is the perfect confirmation flag.

### Initialize player data

> First thing we do is find the program address for the player account using the seed string "player" and the players public key. Then we call initPlayer to create the account.

Reviewer: ```suggestion
First thing we do is find the program address for the player account using the seed string `player"`and the player's public key. Then we call `initPlaye`r to create the account.
```

```js
const [pda] = PublicKey.findProgramAddressSync(
  [Buffer.from("player", "utf8"), publicKey.toBuffer()],
  new PublicKey(LUMBERJACK_PROGRAM_ID)
);

const transaction = program.methods
  .initPlayer()
  .accounts({
    player: pda,
    signer: publicKey,
    systemProgram: SystemProgram.programId,
  })
  .transaction();

const tx = await transaction;
const txSig = await sendTransaction(tx, connection, {
  skipPreflight: true,
});

await connection.confirmTransaction(txSig, "confirmed");

```

### Subscribe to account updates

> Here you can see how to get account data in the js client and how to subscribe to an account. "connection.onAccountChange" creates a socket connection to the RPC node which will push any changes that happen to the account to the client.

Reviewer: ```suggestion
Here you can see how to get account data in the js client and how to subscribe to an account. `connection.onAccountChange` creates a socket connection to the RPC node which will push any changes that happen to the account to the client.
```

> This is way faster than fetching new account data after every change.

Reviewer: ```suggestion
This is faster than fetching new account data after every change.
```

> We can then use the "program.coder" to decode the account data into the TS types and directly use it in the game.

Reviewer: ```suggestion
We can then use the `program.coder` to decode the account data into the TS types and directly use it in the game.
```
